xqproxy.man
.\" Copyright (c) 2009 Vic Lee
.\"
.\" Permission is hereby granted, free of charge, to any person
.\" obtaining a copy of this software and associated documentation
.\" files (the "Software"), to deal in the Software without
.\" restriction, including without limitation the rights to use,
.\" copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the
.\" Software is furnished to do so, subject to the following
.\" conditions:
.\"
.\" The above copyright notice and this permission notice shall be
.\" included in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
.\" EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
.\" OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
.\" NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
.\" HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
.\" WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH XQPROXY 1 __xorgversion__
.SH NAME
xqproxy \- XDMCP Query Proxy
.SH SYNOPSIS
.B xqproxy
[-display n] [-host hostname] [-port port] [-timeout seconds] [-query [-xauth]]
[-manage [-sessionid session-id]]
.SH DESCRIPTION
.I xqproxy
is a proxy program that can send XDMCP query to a display manager to establish
XDM session for an existing local X display.
.SH OPTIONS
.PP
.TP 8
.B \-display n
The local display number. Default is 0.
.PP
.TP 8
.B \-host hostname
The hostname of the remote display manager. Default is localhost.
.PP
.TP 8
.B \-port port
The XDMCP UDP port of the remote display manager. Default is 177.
.PP
.TP 8
.B \-timeout seconds
Timeout value while waiting for response from the server.
.PP
.TP 8
.B \-query
Sends QUERY and REQUEST to the display manager to request a new session. The
session id and magic cookie will be printed in stdout.
.PP
.TP 8
.B \-xauth
When use together with -query option, xqproxy will call xauth to add the magic
cookie to local .Xauthority database.
.PP
.TP 8
.B \-manage
Sends MANAGE to the display manager to start the session. If this option is
used together with -query, it will always use the session id returned from the
display manager. Otherwise, it will uses the session id assigned by -sessionid
argument.
.PP
.TP 8
.B \-sessionid session-id
The new session id returned from the display manager. This option is useful
only if -manage option is used separately.
.SH EXAMPLES
.PP
The following command request the display manager on REMOTEHOST to establish
an XDM session directly to a local X server on display :1 :
.PP
	xqproxy -display 1 -host \fIREMOTEHOST\fP -query -xauth -manage
.PP
The following shell script request the display manager supplied by the first
argument to establish an XDM session over a secure SSH tunnel on a local
display supplied by the second argument (note that xqproxy runs on the server)
:
.PP
	#!/bin/sh
.br
	# Usage: xqssh <server> <local_display_number>
.br
	SSH=$(which ssh)
.br
	if [ "x$SSH" = "x" ]; then
.br
	    echo openssh is required in order to use xqssh.
.br
	    exit 1
.br
	fi
.br
	QUERY=$(ssh $1 xqproxy -display $2 -query)
.br
	echo "$QUERY"
.br
	SESSIONID=$(echo "$QUERY" | awk '/ACCEPT/ {print substr($3,12)}')
.br
	COOKIE=$(echo "$QUERY" | awk '/ACCEPT/ {print substr($7,12,32)}')
.br
	if [ "x$SESSIONID" = "x" ] || [ "x$COOKIE" = "x" ]; then
.br
	    exit 1
.br
	fi
.br
	xauth add :$2 MIT-MAGIC-COOKIE-1 $COOKIE
.br
	PORT=$(expr 6000 + $2)
.br
	ssh -R $PORT:localhost:$PORT -fN $1
.br
	ssh $1 xqproxy -display $2 -manage -sessionid $SESSIONID
.br
	exit 0
.PP
.SH "SEE ALSO"
ssh(1)
.SH BUGS
.PP
Please send bugs to Vic Lee <llyzs@163.com>
.SH AUTHORS
Vic Lee



xqproxy.c
/*
 * Copyright (c) 2009 Vic Lee
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "xdmcpclient.h"

#define XQPROXY_ERROR_ARGUMENT  1
#define XQPROXY_ERROR_INIT      2
#define XQPROXY_ERROR_RESOLVE   3
#define XQPROXY_ERROR_QUERY     4
#define XQPROXY_ERROR_REQUEST   5
#define XQPROXY_ERROR_XAUTH     6
#define XQPROXY_ERROR_MANAGE    7

int
main (int argc, char *argv[])
{
    XdmcpClient *client;
    int ret = 0;
    int i;
    int display = 0;
    unsigned int sessionid = 0;
    const char *host = "localhost", *port = "177";
    int do_query = 0;
    int do_xauth = 0;
    int do_manage = 0;
    int timeout = 5;

    for (i = 1; i < argc; i++)
    {
        if (strcmp (argv[i], "-display") == 0)
        {
            if (i + 1 >= argc)
            {
                printf ("Error> Missing argument value %s\n", argv[i]);
                return XQPROXY_ERROR_ARGUMENT;
            }
            display = atoi (argv[++i]);
            continue;
        }
        if (strcmp (argv[i], "-host") == 0)
        {
            if (i + 1 >= argc)
            {
                printf ("Error> Missing argument value %s\n", argv[i]);
                return XQPROXY_ERROR_ARGUMENT;
            }
            host = argv[++i];
            continue;
        }
        if (strcmp (argv[i], "-port") == 0)
        {
            if (i + 1 >= argc)
            {
                printf ("Error> Missing argument value %s\n", argv[i]);
                return XQPROXY_ERROR_ARGUMENT;
            }
            port = argv[++i];
            continue;
        }
        if (strcmp (argv[i], "-sessionid") == 0)
        {
            if (i + 1 >= argc)
            {
                printf ("Error> Missing argument value %s\n", argv[i]);
                return XQPROXY_ERROR_ARGUMENT;
            }
            sessionid = (unsigned int) atoi (argv[++i]);
            continue;
        }
        if (strcmp (argv[i], "-timeout") == 0)
        {
            if (i + 1 >= argc)
            {
                printf ("Error> Missing argument value %s\n", argv[i]);
                return XQPROXY_ERROR_ARGUMENT;
            }
            timeout = (unsigned int) atoi (argv[++i]);
            continue;
        }
        if (strcmp (argv[i], "-query") == 0)
        {
            do_query = 1;
            continue;
        }
        if (strcmp (argv[i], "-xauth") == 0)
        {
            do_xauth = 1;
            continue;
        }
        if (strcmp (argv[i], "-manage") == 0)
        {
            do_manage = 1;
            continue;
        }
    }

    if (do_xauth && !do_query)
    {
        printf ("Error> Argument -xauth must be used together with -query\n");
        return XQPROXY_ERROR_ARGUMENT;
    }

    if (!do_query && !do_manage)
    {
        printf ("Error> Either -query or -manage must be supplied\n");
        return XQPROXY_ERROR_ARGUMENT;
    }

    if (do_manage && !sessionid && !do_query)
    {
        printf ("Error> Argument -sessionid must be supplied if -manage is used without -query\n");
        return XQPROXY_ERROR_ARGUMENT;
    }

    if (sessionid && do_query)
    {
        printf ("Error> Argument -sessionid cannot be used together with -query\n");
        return XQPROXY_ERROR_ARGUMENT;
    }

    client = XdmcpClientNew (display, timeout);
    if (!client) return XQPROXY_ERROR_INIT;
    if (XdmcpClientRegisterServer (client, host, port))
    {
        if (do_query)
        {
            if (!XdmcpClientQuery (client)) ret = XQPROXY_ERROR_QUERY;
            else if (!XdmcpClientRequest (client)) ret = XQPROXY_ERROR_REQUEST;
            else if (do_xauth && !XdmcpClientAddXauth (client)) ret = XQPROXY_ERROR_XAUTH;
        }
        if (ret == 0 && do_manage)
        {
            if (!XdmcpClientManage (client, sessionid))
            {
                ret = XQPROXY_ERROR_MANAGE;
            }
        }
    }
    else
    {
        ret = XQPROXY_ERROR_RESOLVE;
    }
    XdmcpClientFree (client);
    return ret;
}
